### ISSUE TITLE
http 500 no longer triggering http errors

### ISSUE THREAD

**Issue by Dieterbe** [raintank/grafana#372]
I have an endpoint that in 80% of the time, returns a http 500 and the text 'panic' instead of 'ok' which we match on
```
dieter@dieter-m6800 ~ for i in $(seq 1 100); do curl -I -s http://metrics20.org:8888/bad/80 | grep '^HTTP'; done | sort | uniq -c
     20 HTTP/1.1 200 OK
     80 HTTP/1.1 500 Internal Server Error
```
I updated my endpoint https://app.raintank.io/endpoints/summary/152 today at 10:12 UTC
however, the http uptime and http status just stays at 100%

i'm pretty sure this used to work fine.
@nopzor1200 @woodsaj 

PS i recommend you also use the [bad-http-server](https://github.com/raintank/bad-http-server) instance running at http://metrics20.org:8888/ to verify some our logic

**Comment by woodsaj** [comment]
we dont currently use response code for determining if an error is present.  We used to have a config paramater for valideStatusCode, but that was removed a long time ago in favor of just storing the statusCode as a metric and allowing users to to set thresholds on it (we used to be able to do this with realtime alerting)

The only time an error will be raised is if we cant connect, or if the content-match fails. 

I just added content-match of `ok` to the check configuration and it is now reporting the site as down most of the time.

**Comment by Dieterbe** [comment]
ooh ok so i must have just forgotten to specify the content-match and mistakenly presumed it would use the http code. i'll verify this a bit more and can probably just close this.

**Comment by nopzor1200** [comment]
This is kind of similar to the change we made where we set error state if packetloss = 100, imo.

What do you think of a similar idea to set error state on the test if the statuscode returned >= 400.

I think it's a reasonable default to make for now, until we have more sophisticated alerting.

(More sophisticated alerting would let you define what statuscodes for http and/or what levels of packetloss for ping would cause an error state for a test, etc etc)

But making that change (for the collector to set error state if the http response code is >400) seems like it'd cause a better out of the box experience in the meantime, just like we did with packetloss?

Thoughts?

**Comment by Dieterbe** [comment]
I like having >= 400 be treated as errorous by default, unless we provide the knobs. but I'm not convinced this decision should be moved into the collector.  it seems more flexible and simpler to implement that decision in the receiving end of the metrics, and keep this concern away from the probe/collector. i like @woodsaj's approach of storing the http code as a metric, this gives us a lot of flexibility in doing queries on it and doing all kinds of decisions based on what the val is, leveraging the alerting expressions.  so how about we extend the alerting expression to include an `<if threshold of error states met> OR if status code >= 400` (for given number of collectors, x steps in a row). this does imply we track http code on a per-collector basis of course.

**Comment by nopzor1200** [comment]
It'll be much cheaper for the collectors to set error state, similar to
how they do for packet loss right now.

Regardless of that detail - guess you and I agree on the concept of
"erroneous by default" for >400

-r

On Wednesday, July 29, 2015, Dieter Plaetinck <notifications@github.com>
wrote:

> I like having >= 400 be treated as errorous by default, unless we provide
> the knobs. but I'm not convinced this decision should be moved into the
> collector. it seems more flexible and simpler to implement that decision in
> the receiving end of the metrics, and keep this concern away from the
> probe/collector. i like @woodsaj <https://github.com/woodsaj>'s approach
> of storing the http code as a metric, this gives us a lot of flexibility in
> doing queries on it and doing all kinds of decisions based on what the val
> is, leveraging the alerting expressions. so how about we extend the
> alerting expression to include an <if threshold of error states met> OR
> if status code >= 400 (for given number of collectors, x steps in a row).
> this does imply we track http code on a per-collector basis of course.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/raintank/grafana/issues/372#issuecomment-125940325>.
>


-- 
Raj Dutt

co-founder/ceo, raintank
+1-917-664-1310
dutt@raintank.io

**Comment by nopzor1200** [comment]
Oh and just like packet loss I agree storing it as a metric still no
matter what we do. Once we have more alerting capabilities that will be
super useful.

On Wednesday, July 29, 2015, Raj Dutt <dutt@raintank.io> wrote:

> It'll be much cheaper for the collectors to set error state, similar to
> how they do for packet loss right now.
>
> Regardless of that detail - guess you and I agree on the concept of
> "erroneous by default" for >400
>
> -r
>
> On Wednesday, July 29, 2015, Dieter Plaetinck <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> I like having >= 400 be treated as errorous by default, unless we provide
>> the knobs. but I'm not convinced this decision should be moved into the
>> collector. it seems more flexible and simpler to implement that decision in
>> the receiving end of the metrics, and keep this concern away from the
>> probe/collector. i like @woodsaj <https://github.com/woodsaj>'s approach
>> of storing the http code as a metric, this gives us a lot of flexibility in
>> doing queries on it and doing all kinds of decisions based on what the val
>> is, leveraging the alerting expressions. so how about we extend the
>> alerting expression to include an <if threshold of error states met> OR
>> if status code >= 400 (for given number of collectors, x steps in a
>> row). this does imply we track http code on a per-collector basis of course.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/raintank/grafana/issues/372#issuecomment-125940325>.
>>
>
>
> --
> Raj Dutt
>
> co-founder/ceo, raintank
> +1-917-664-1310
> dutt@raintank.io <javascript:_e(%7B%7D,'cvml','dutt@raintank.io');>
>
>

-- 
Raj Dutt

co-founder/ceo, raintank
+1-917-664-1310
dutt@raintank.io

**Comment by Dieterbe** [comment]
played a bit with it more. was indeed simply the content match setting. only thing left to do is implement the error on >= 400. @woodsaj can you check both suggested approaches above and tell us if nopzor's approach looks good. i'm ok with nopzor's approach (report http code as metric as we do, but have collector set error state on >= 400), i'll definitely be simpler then doing it as part of alerting execution.

**Comment by woodsaj** [comment]
Adding a `if statusCode >= 400` condition is only a few lines of code and can easily be done.

@Dieterbe @nopzor1200 
My only question here is if we want to allow the user to specify the set of valid http codes?

>= 400 makes sense, but some users may also consider 300 codes to also be an error.

**Comment by nopzor1200** [comment]
@woodsaj I agree some may consider 300 code to be an error, just like some users may consider 50% packetloss to be an "error". 

We can cater to users who consider 300 to be an error (or anything other than a 200, or anything other than a particular 301 that they're testing, who knows) with more "generic" alerting (I hope we can kick off a discussion about the scope of v1 next week). No need to clutter up our UI for now.

I think its safe to say that _all_ users will consider >400 to be an error for any http/https "endpoint",  just like _all_ users will consider 100% packetloss to be an "error" on any ping "endpoint".

If you agree with the logic lets make this change. Glad it's a quick one.

**Comment by Dieterbe** [comment]
+1 on >= 400 as sensible default, we can worry about user driven customizations later.

**Comment by woodsaj** [comment]
i have opened https://github.com/raintank/raintank-collector/issues/16 for this.

**Comment by Dieterbe** [comment]
closing this then.
